### DISCUSSION TITLE
Problem with average speed

### DISCUSSION THREAD

**Discussion by DolomitesTom** [devemux86/cruiser#289] [Q&A] [answered]
Today I had a problem with the average speed. At first stop I had 72 av.speed, one colleague with Calimoto 69, BMW app 68. Then I paused the navigation, but didn't quit. The second stage was quicker. The av.speed from the colleagues went to 86 (Calimoto) and 84 (BMW), mine with Cruiser was frozen at 73 and did not update any more. We had 3 more pauses were I only paused the app. It seems that the av.speed is not updating any more after pausing.

**Comment by devemux86** [comment]
@DolomitesTom thanks for the report.

I'm trying to reproduce it but can't so far.

Do you remember more details of your actions?
Did you just pause navigation and resume later or anything else?

---

Have other users seen any issue with the average speed?

**Reply by DolomitesTom** [reply]
In this group I was the only one with Cruiser, some use Calimoto, some the BMW app and some nothing, they just rely on my guiding and are firing behind.

I did only pressed on pause navigation, sent the app into the background and then resumed navigation after the pause. Since the overview is still not very comfortable, I often pause navigation to do a zoom to get the next 50 kilometers. 
I did a proposal for the arrow button to make separate zoom levels of the route, this would help a lot.

I also had one issue after a pause where I deleted just the nearest stop (WP) and suddenly it calculated the route 3 hours longer as normal. The reason for this was that the last 12 shaping points have not been recognized as already gone. I had to stop navigation and delete manually the last 12 SPs. 

I am suffering to the point where average speed is availabe as history value with the recorded track some day. Also when the app is closed, the average speed should not be initialized but saved somehow and then is continuously updated when the tours is continued.

**Reply by devemux86** [reply]
> The reason for this was that the last 12 shaping points have not been recognized as already gone.

Do you mean the points you had passed and were behind you?
Navigation recognizes them in all rerouting actions, skip, etc.

When we **pause** the navigation, we return to route planning where the route has all waypoints.
There is room for improvement here, at least in paused navigation, I will see what can be done. :slightly_smiling_face: 

> Also when the app is closed, the average speed should not be initialized but saved somehow and then is continuously updated when the tours is continued.

This is a good point, I will put it in my list.

**Reply by DolomitesTom** [reply]
> Do you mean the points you had passed and were behind you?

yes. It didn't recognize it somehow that these 12 SPs were passed. Yes, maybe this pausing is the root cause.

**Reply by devemux86** [reply]
It's because with pause navigation we return to normal route planning where everything is available.

But it should not matter, when we resume navigation it starts anywhere on the new calculated route.

**Reply by DolomitesTom** [reply]
With strict routing it is a problem, when the last 12 shaping points are included in the new calculation, because then it leads me back to where I came from.

**Reply by devemux86** [reply]
It does not happen, already tested it:

- Strict navigation on
- Pass some waypoints
- Pause navigation
- Change the route
- Resume navigation
- Navigation restarts anywhere on the route
(from there route must be strictly followed)

**Comment by 0709wiwiwi** [comment]
@DolomitesTom

Can it be that some displays do show the current trkseg avg ? 

Probably better to attach the track record here ?
For a first analysis load into the Android gpx viewer.
Note the avg speed and the avg speed @ movement

I had a very short track record test. 
Trk with 3 trkseg by 2 x navpause + end.

Data as read by gpx editor (windows)

Opening 20230821_185643_slow_fast_slow.gpx
0 route points
0 routes (0 points))
1 tracks (151 points in 3 track segments)
TGPX
Track #1 : ( 3 track segments)
Track segment #1 (49 points)
 778.3 m / 0:04:43 = 9.9 km/h
Track segment #2 (68 points)
 2.2 km / 0:06:09 = 21.2 km/h
Segment 3 (34 points)
 517.2 m / 0:03:20 = 9.3 km/h

**Average values shown in the cruiser app:**
- After trkseg 1: 10 km/h.  
By Android gpx viewer > 9.9 kmh avg and 9.9 kmh avg movement.
- After trkseg 1+2: 16 km/h. 
By Android gpx viewer > 14.4 avg and 16.3 kmh avg movement.
- After trkseg 1+2+3: 14 km/h. 
By Android gpx viewer > 12.2 kmh avg and 14.6 kmh avg movement.

**Comment by DolomitesTom** [comment]
Do you mean to load the track gpx into a 3rd party app called "Android GPX viewer"? I have never used it and wouldn't, after reading the data privacy guidelines:

This app may share the following types of data with third parties.
Location, app activity, and 2 others

This app may collect the following types of data
Personal data and app activities

Then again, it's not that important. I hope that devemux still expands the statistics

**Reply by 0709wiwiwi** [reply]
If I'm reading it right, it's the developer providing this general info. Whether or not this is than complete or correct, I don't know. However, as soon as you have a mobile phone or smartphone in your hands, privacy is an issue. No subject to dwell on for too long indeed, so just let's continue playing with GPS.

**Reply by devemux86** [reply]
> it's the developer providing this general info

Google first approves this information by "reviewing" the app.

**Comment by devemux86** [comment]
Reminder:

_Strict navigation is currently not allowed to start on the last segment of the route._
(after the last via / shaping point and before destination)

This is to avoid complications with round trips ending and starting on same road.
Perhaps @DolomitesTom had a similar case, all other cases should work, see https://github.com/devemux86/cruiser/discussions/289#discussioncomment-6779810.

We made this decision a long time ago, not sure if it's the right one or it should be optional.

**Reply by 0709wiwiwi** [reply]
There was once (Kurviger) a very brief discussion about this.  
Like place an extra shaping point very shortly before the end. 
A forgotten topic, so probably won't be that important either ?

**Comment by DolomitesTom** [comment]
> Perhaps @DolomitesTom had a similar case, all other cases should work

Unfortunately not, I still had 15 SPs to go.

This was the tour. We were at WP 25, but not exactly at 25, some meters apart, you see it on the purple track.

<img src="https://github.com/devemux86/cruiser/assets/121371629/7604103a-3840-43fe-9b03-67ee5b4594e3" width="50%"/><img src="https://github.com/devemux86/cruiser/assets/121371629/e07ea672-3095-42ec-9518-5fdb1ee2ce5b" width="50%"/>

Then I resumed the navigation and deleted WP25. Then the tour was recalculated to SP 14. It assumed obviously that we had passed already WP 1 to SP 13. But we had already passed all till SP 24. In principle also till WP 25, but as told, not exactly, since this was not on the road. Therefore I deleted WP 25 in order to not get routed back to WP25 when continuing the tour. And then the wrong calculation occurred.

**Reply by devemux86** [reply]
@DolomitesTom thanks for the details.

Can you send me the route and the app settings to check it?

**Reply by devemux86** [reply]
> Then I resumed the navigation and deleted WP25.

I also do not understand clearly this part, what were you trying to do?
Why not simply return to the route and continue navigation?

When delete one waypoint the route is calculated again with all remaining waypoints.
All waypoints before the deleted waypoint + all waypoints after the deleted waypoint.

> Then the tour was recalculated to SP 14.

So based on my above description, what does this mean?

It is not a rerouting _by navigation_ to use the user location + the remaining waypoints.
Manually deleting waypoints or changing the route is route planning with all waypoints.

**Reply by devemux86** [reply]
An alternative way is to use the navigation action to skip many waypoints.
Click it and select the next waypoint to continue navigation from user location.

**Reply by DolomitesTom** [reply]
not delete wps during the tour after the pause would have been better or skip all like advised, but sometimes you don't think to the easiest solutions when beeing in the middle of a tour and the sun burns to the head :-)

**Reply by devemux86** [reply]
@DolomitesTom thanks for the route. I made the following test:

- Strict navigation on
- Import the route
- Start navigation, go to via point 25
- Pause / resume navigation
- Delete via point 25 (with long press or waypoints list)
- Whole route is calculated with all remaining waypoints
- When we return on the route, navigation continues
(and from there route must be strictly followed)

To avoid the whole route calculation we must use the navigation actions that provide **partial routing**.
Select as destination the waypoint 26 and a route will be calculated from user location to waypoint 26,
while the route remains unchanged.

**Comment by DolomitesTom** [comment]
> Select as destination the waypoint 26 

This is something I don't understand. Do you mean I should have set the WP 26 as last WP of the tour?

My understanding is I should not have deleted the waypoint 25, shouldn't I?

**Reply by devemux86** [reply]
Indeed why you deleted a waypoint and not just returned to the route to continue the navigation?

Even if you wanted to change the route during navigation, it is better to use the navigation tools.

Remember: route planning calculates the whole route. This will change in the future:
- https://github.com/devemux86/cruiser/discussions/142

Navigation tools (all of them) calculate only until the next waypoint, so the _route does not change_.

For example:
- Use the 2nd action "Destination" with the "?". Select waypoint 26 to calculate a route until 26.
- Or if you were off the route, tap the warning symbol at top to calculate until next waypoint 26.
- Even if you stopped navigation, the rerouting symbol asks you on what waypoint to calculate.

<img src="https://github.com/devemux86/cruiser/assets/3484020/2bc56514-8b87-4798-8789-79c12764a060" width="50%"/><img src="https://github.com/devemux86/cruiser/assets/3484020/b57cdeb1-45f4-4463-80c9-f1cf463500fa" width="50%"/>

**Reply by DolomitesTom** [reply]
Afterwards you are always wiser 😂

**Comment by devemux86** [comment]
Average speed will be saved / restored in next Cruiser.
So you can exit the app without losing any statistics.

It will also be part of the new profiles. Each app profile can measure and store its own average speed.

- https://github.com/devemux86/cruiser/discussions/165#discussioncomment-6800089

**Reply by DolomitesTom** [reply]
Thanks a lot

**Reply by devemux86** [reply]
> Average speed will be saved / restored in next Cruiser.

@DolomitesTom you can test it in:
- https://github.com/devemux86/cruiser/discussions/299

**Comment by devemux86** [comment, marked as answer]
> Average speed will be saved / restored in next Cruiser.

Implemented in [Cruiser 3.0.32](https://play.google.com/store/apps/details?id=gr.talent.cruiser).
